src/ring_theory/perfection.lean

```diff
@@ -0,0 +1,422 @@
+`{ f : ℕ → M | ∀ n, f (n + 1) ^ p = f n }`. -/
+def submonoid.perfection (M : Type u) [comm_monoid M] (p : ℕ) : submonoid (ℕ → M) :=
+{ carrier := { f | ∀ n, f (n + 1) ^ p = f n },
+  one_mem' := λ n, one_pow _,
+  mul_mem' := λ f g hf hg n, (mul_pow _ _ _).trans $ congr_arg2 _ (hf n) (hg n) }
+
+/-- The perfection of a semiring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def perfection_s (R : Type u) [comm_semiring R] (p : ℕ) [hp : fact p.prime] [char_p R p] :
+  subsemiring (ℕ → R) :=
+{ zero_mem' := λ n, zero_pow $ hp.pos,
+  add_mem' := λ f g hf hg n, (frobenius_add R p _ _).trans $ congr_arg2 _ (hf n) (hg n),
+  .. perfection_m R p }
+  
+/-- The perfection of a ring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def perfection (R : Type u) [comm_ring R] (p : ℕ) [hp : fact p.prime] [char_p R p] :
+  subring (ℕ → R) :=
+{ neg_mem' := λ f hf n, (frobenius_neg R p _).trans $ congr_arg _ (hf n),
+  .. perfection_s R p }
+
+namespace ring_char
+
+instance char_p (R : Type u) [semiring R] : char_p R (ring_char R) :=
+⟨spec R⟩
+
+theorem of_eq {R : Type u} [semiring R] {p : ℕ} (h : ring_char R = p) : char_p R p :=
+h ▸ ring_char.char_p R
+
+theorem eq_iff {R : Type u} [semiring R] {p : ℕ} : ring_char R = p ↔ char_p R p :=
+⟨of_eq, eq.symm ∘ eq R⟩
+
+theorem dvd {R : Type u} [semiring R] {x : ℕ} (hx : (x : R) = 0) : ring_char R ∣ x :=
+(spec R x).1 hx
+
+end ring_char
+
+namespace char_p
+
+instance pi (ι : Type u) [hi : nonempty ι] (R : Type v) [semiring R] (p : ℕ) [char_p R p] :
+  char_p (ι → R) p :=
+⟨λ x, let ⟨i⟩ := hi in iff.symm $ (char_p.cast_eq_zero_iff R p x).symm.trans
+⟨λ h, funext $ λ j, show ring_hom.apply (λ _, R) j (↑x : ι → R) = 0,
+    by rw [ring_hom.map_nat_cast, h],
+  λ h, (ring_hom.apply (λ _, R) i).map_nat_cast x ▸ by rw [h, ring_hom.map_zero]⟩⟩
+
+-- diamonds
+instance pi' (ι : Type u) [hi : nonempty ι] (R : Type v) [comm_ring R] (p : ℕ) [char_p R p] :
+  char_p (ι → R) p :=
+char_p.pi ι R p
+
+instance subsemiring (R : Type u) [semiring R] (p : ℕ) [char_p R p] (S : subsemiring R) :
+  char_p S p :=
+⟨λ x, iff.symm $ (char_p.cast_eq_zero_iff R p x).symm.trans
+⟨λ h, subtype.eq $ show S.subtype x = 0, by rw [ring_hom.map_nat_cast, h],
+  λ h, S.subtype.map_nat_cast x ▸ by rw [h, ring_hom.map_zero]⟩⟩
+
+instance subring (R : Type u) [ring R] (p : ℕ) [char_p R p] (S : subring R) :
+  char_p S p :=
+⟨λ x, iff.symm $ (char_p.cast_eq_zero_iff R p x).symm.trans
+⟨λ h, subtype.eq $ show S.subtype x = 0, by rw [ring_hom.map_nat_cast, h],
+  λ h, S.subtype.map_nat_cast x ▸ by rw [h, ring_hom.map_zero]⟩⟩
+
+instance subring' (R : Type u) [comm_ring R] (p : ℕ) [char_p R p] (S : subring R) :
```

Comment:
Can't TC inference find this from the instance above it? That would be bad.

Reply:
diamonds

Reply:
Hmm, that's bad. Can we fix the diamonds?

Reply:
I don't believe it's inherently bad. Rather, this can be treated as a typeclass shortcut to make Lean's work easier.